orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,206 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
   graph.AddInitializedTensor(proto_data);
 }
 
+Status FindAllConnectedNodes(Graph& graph,
+                             const Node* node,
```

Comment:
Can this be "const Node &" to avoid the possibility of null?  Otherwise, could you add an assert that node is non-null.

Reply:
The caller of this method checks to ensure it is not null. But I'll add an assert in the function in case it is being used somewhere else later without any caller check. 

Thinking about it a bit more, maybe it is not a bad idea to adapt some annotation in the code. Annotations like _in_, _out_, _not_null_ will help people better understand the function and less likely to use it wrong. Plus, compiler can also provide code analysis based on that. Microsoft SAL is an example. ([https://docs.microsoft.com/en-us/cpp/code-quality/using-sal-annotations-to-reduce-c-cpp-code-defects?view=vs-2019](https://docs.microsoft.com/en-us/cpp/code-quality/using-sal-annotations-to-reduce-c-cpp-code-defects?view=vs-2019)). I feel this could be of more help especially in the public API when facing external users. But SAL is only supported in MSVC. I am not sure if other compilers provide something similar... 

---
In reply to: [452253970](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r452253970) [](ancestors = 452253970)

Reply:
More checking generally would be a great direction, but as you say there is not much standardization.   We could probably make more use of C++ static_assert than we are right now (e.g. there was a bug I wrote a few weeks back by adding an object without a trivial destructor to a thread-local, which was a combination not supported on one of the platforms we build on -- adding a static assert would have caught that on my dev box rather than CI).  We could probably also use dynamic tools like Valgrind to help pick up some common bugs.